repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.powerfilter
+
+import java.io.IOException
+import java.net.{MalformedURLException, URL}
+import java.util.Optional
+import java.util.concurrent.atomic.AtomicReference
+
+import com.codahale.metrics.MetricRegistry
+import com.typesafe.scalalogging.slf4j.StrictLogging
+import javax.inject.{Inject, Named}
+import javax.servlet.http.HttpServletResponse.{SC_INTERNAL_SERVER_ERROR, SC_REQUEST_TIMEOUT}
+import javax.servlet.http.{HttpServlet, HttpServletRequest, HttpServletResponse}
+import javax.servlet.{RequestDispatcher, ServletContext}
+import org.apache.commons.lang3.StringUtils
+import org.openrepose.commons.config.manager.UpdateListener
+import org.openrepose.commons.utils.StringUriUtilities
+import org.openrepose.commons.utils.http.CommonRequestAttributes
+import org.openrepose.commons.utils.io.stream.ReadLimitReachedException
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, RouteDestination}
+import org.openrepose.core.domain.Port
+import org.openrepose.core.filter.SystemModelInterrogator
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.services.reporting.ReportingService
+import org.openrepose.core.services.reporting.metrics.MetricsService
+import org.openrepose.core.services.routing.robin.Clusters
+import org.openrepose.core.spring.ReposeSpringProperties
+import org.openrepose.core.systemmodel.config._
+import org.openrepose.nodeservice.request.RequestHeaderService
+import org.openrepose.nodeservice.response.ResponseHeaderService
+import org.openrepose.powerfilter.ReposeRoutingServlet._
+import org.springframework.beans.factory.annotation.Value
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+@Named
+class ReposeRoutingServlet @Inject()(@Value(ReposeSpringProperties.NODE.CLUSTER_ID) clusterId: String,
+                                     @Value(ReposeSpringProperties.NODE.NODE_ID) nodeId: String,
+                                     configurationService: ConfigurationService,
+                                     requestHeaderService: RequestHeaderService,
+                                     responseHeaderService: ResponseHeaderService,
+                                     reportingService: ReportingService,
+                                     optMetricsService: Optional[MetricsService]
+                                    ) extends HttpServlet with UpdateListener[SystemModel] with StrictLogging {
+
+  private val localCluster = new AtomicReference[Option[ReposeCluster]](None)
```

Comment:
Why all the atomic references? Especially on things that have no ability to change?

Reply:
I'm pretty sure the metrics service can't change, it's flagged optional here because we are sort of heading that way, but any move along those lines will involve this entire context getting remade. You didn't really explain why were are using atomic references here, versus most places? Also at least half the things being referenced would cause the node to be torn down and rebuilt from scratch which would replace the whole servlet.